src/Compilers/CSharp/Portable/Syntax/CSharpLineDirectiveMap.cs

```diff
@@ -21,69 +21,143 @@ public CSharpLineDirectiveMap(SyntaxTree syntaxTree)
         // Add all active #line directives under trivia into the list, in source code order.
         protected override bool ShouldAddDirective(DirectiveTriviaSyntax directive)
         {
-            return directive.IsActive && directive.Kind() == SyntaxKind.LineDirectiveTrivia;
+            return directive.IsActive && (directive.Kind() is SyntaxKind.LineDirectiveTrivia or SyntaxKind.LineSpanDirectiveTrivia);
         }
 
         // Given a directive and the previous entry, create a new entry.
         protected override LineMappingEntry GetEntry(DirectiveTriviaSyntax directiveNode, SourceText sourceText, LineMappingEntry previous)
         {
             Debug.Assert(ShouldAddDirective(directiveNode));
-            var directive = (LineDirectiveTriviaSyntax)directiveNode;
+            var directive = (LineOrSpanDirectiveTriviaSyntax)directiveNode;
 
             // Get line number of NEXT line, hence the +1.
             var directiveLineNumber = sourceText.Lines.IndexOf(directive.SpanStart) + 1;
 
             // The default for the current entry does the same thing as the previous entry, except
             // resetting hidden.
             var unmappedLine = directiveLineNumber;
-            var mappedLine = previous.MappedLine + directiveLineNumber - previous.UnmappedLine;
-            var mappedPathOpt = previous.MappedPathOpt;
-            PositionState state = PositionState.Unmapped;
 
             // Modify the current entry based on the directive.
-            SyntaxToken lineToken = directive.Line;
-            if (!lineToken.IsMissing)
+            switch (directive)
             {
-                switch (lineToken.Kind())
-                {
-                    case SyntaxKind.HiddenKeyword:
-                        state = PositionState.Hidden;
-                        break;
-
-                    case SyntaxKind.DefaultKeyword:
-                        mappedLine = unmappedLine;
-                        mappedPathOpt = null;
-                        state = PositionState.Unmapped;
-                        break;
-
-                    case SyntaxKind.NumericLiteralToken:
-                        // skip both the mapped line and the filename if the line number is not valid
-                        if (!lineToken.ContainsDiagnostics)
+                case LineDirectiveTriviaSyntax lineDirective:
+                    {
+                        var mappedLine = (previous.State == PositionState.RemappedSpan) ?
+                            unmappedLine :
```

Comment:
I didn't understand this. It seems strange to use something from "unmapped" domain directly in "mapped" domain without any sort of mapping.